### ISSUE TITLE
How to run it

### ISSUE THREAD

**Issue by vaskomichal** [pzmarzly/ancs4linux#2]
Hi, I try to run this program on my raspberry pi zero W. I installed dbus-python - pip install dbus-python. Paired my iPhone with raspberry and connected it. 

Try to run script ./main.py - the script is running but I don't see any info on the console nothing happened when the notification arrived.

Thanks

**Comment by pzmarzly** [comment]
Hi. Thanks for being interested in the project and sorry for replying late.

I think that you paired your devices in such a way that iPhone doesn't think your Raspberry is capable of receiving notifications. I have no idea why this happens, but I personally tried 2 different laptops and on each of them different pairing methods worked or did not. The one that I found to be most likely to work is the one I described in [README.md#Pairing](https://github.com/pzmarzly/ancs4linux/blob/master/README.md#pairing), but I never checked it on any Raspberry.

Basically, if the script is not printing anything, then it is most likely stuck searching for a connected device that advertises the right services. But there is a possibility that something else went wrong - as you can see, the project is on pretty early stage of development, I hope I'll find time to research Bluetooth more one day.

**Comment by vaskomichal** [comment]
> Hi. Thanks for being interested in the project and sorry for replying late.
> 
> I think that you paired your devices in such a way that iPhone doesn't think your Raspberry is capable of receiving notifications. I have no idea why this happens, but I personally tried 2 different laptops and on each of them different pairing methods worked or did not. The one that I found to be most likely to work is the one I described in [README.md#Pairing](https://github.com/pzmarzly/ancs4linux/blob/master/README.md#pairing), but I never checked it on any Raspberry.
> 
> Basically, if the script is not printing anything, then it is most likely stuck searching for a connected device that advertises the right services. But there is a possibility that something else went wrong - as you can see, the project is on pretty early stage of development, I hope I'll find time to research Bluetooth more one day.

Thank you very much for your response! Do I need to set somewhere my device id? Now I try to write a command which I send manually from the command line I want to see if the connection is correct or not. When I send a command manually I see a notification. But when I try to send a request for more info I have an Invalid parameter (0xA2). But maybe my request is incorrect It is complicated for me to create it. Is it possible to send me an example of a request? Which I try to send from the command line? 

Here is my notification command line response bytearray(b'\x00\x15\t\x01\x00\x00\x00\x00') If I understand correctly it is 
id: 0, event_id: 0, event_flags: 21, category_id: 9, category_name: BusinessAndFinance, category_count: 1

The problem is now how to prepare data for the command-line test.

if I understand correctly first part of the request is 

commandId = 0 for example x00
notification Id = 0 - but need to be 4 byte x00 x00 x00 x00
attribute Id = 0 for example App Identifier - x00

I'm new in this conversion I don't know if is it correct.

request what I send is - bytearray(b'\x00\x00\x00\x00\x00\x00') Is it correct ? or I need to add something other because this request doesn't work.

But It looks that the connection works between the command line from raspberry and iPhone. I try to investigate where is the problem and if I found a solution I'll write you maybe it will be helpful in the future. 

If you can check my request or send me an example of what it should look like, I will be grateful.

**Comment by pzmarzly** [comment]
Thanks for looking into it more.

Basically, from what I remember, the iPhone->PC/smartwatch data flow is something like this:
- PC subscribes to Data Source and Notification Source (done via `StartNotify()` in Device.py)
- iPhone prompts user whether to allow notification sharing
- When notification comes, iPhone sends NS message with type 0 (new notification), 1 (notification edited) or 2 (notification removed)
- ancs4linux replies with list of attributes it wants to fetch (`fetchAttributes` variable, which is currently app id, title and message). This has to be send over CP (Control Plane)
- iPhone replies on DS channel (Data Source) with requested data

You can take a look at Device.py:

https://github.com/pzmarzly/ancs4linux/blob/b6fd734d904372bf6d746c0574b256458f5178e5/Device.py#L43-L74

Line 43 is reading from NS, 54 replying over CP, 58 reading from DS.

All of this is described [in ANCS documentation](https://developer.apple.com/library/archive/documentation/CoreBluetooth/Reference/AppleNotificationCenterServiceSpecification/Specification/Specification.html) though it's hardly an easy reading.

So, I think you are crafting packets correctly, but you need to be careful with there being 3 communication channels (called Characteristics in BLE nomenclature).

**Comment by vaskomichal** [comment]
> Thanks for looking into it more.
> 
> Basically, from what I remember, the iPhone->PC/smartwatch data flow is something like this:
> 
> * PC subscribes to Data Source and Notification Source (done via `StartNotify()` in Device.py)
> * iPhone prompts user whether to allow notification sharing
> * When notification comes, iPhone sends NS message with type 0 (new notification), 1 (notification edited) or 2 (notification removed)
> * ancs4linux replies with list of attributes it wants to fetch (`fetchAttributes` variable, which is currently app id, title and message). This has to be send over CP (Control Plane)
> * iPhone replies on DS channel (Data Source) with requested data
> 
> You can take a look at Device.py:
> 
> https://github.com/pzmarzly/ancs4linux/blob/b6fd734d904372bf6d746c0574b256458f5178e5/Device.py#L43-L74
> 
> Line 43 is reading from NS, 54 replying over CP, 58 reading from DS.
> 
> All of this is described [in ANCS documentation](https://developer.apple.com/library/archive/documentation/CoreBluetooth/Reference/AppleNotificationCenterServiceSpecification/Specification/Specification.html) though it's hardly an easy reading.
> 
> So, I think you are crafting packets correctly, but you need to be careful with there being 3 communication channels (called Characteristics in BLE nomenclature).

Hi, I looked at it again. 
My responses from Hci
1. Children return ['dev_34_42_62_6C_0F_C7', 'dev_F8_2D_7C_95_18_6E']   ->  `self.all_children(hci, "dev_")`
2. Connected = 1  ->  `connected = props.Get("org.bluez.Device1", "Connected")`
3. Children return [] -> `self.all_children(device, "service")`
     
     Inside all_children function
     3.a   Children [<DOM Element: node at 0xb61a73f0>, <DOM Element: node at 0xb61a7440>, <DOM Element: node at 0xb61a7490>]

Do you have an idea why I don't have any services inside the all_children function?

Thanks!

**Comment by pzmarzly** [comment]
Both devices seem to be your iPhone, and I think one of it is "regular" Bluetooth (Bluetooth 2.0/3.0) and other is Bluetooth Low Energy, which is what we are looking for.

Now, the problem is depending on the method that was used to pair devices, iPhone sometimes doesn't advertise its services. I don't know why that happens yet. In my experience, pairing via nRF Connect app, while advertising as heartrate meter (via bluez demo Python script), works the best.

**Comment by vaskomichal** [comment]
> Both devices seem to be your iPhone, and I think one of it is "regular" Bluetooth (Bluetooth 2.0/3.0) and other is Bluetooth Low Energy, which is what we are looking for.
> 
> Now, the problem is depending on the method that was used to pair devices, iPhone sometimes doesn't advertise its services. I don't know why that happens yet. In my experience, pairing via nRF Connect app, while advertising as heartrate meter (via bluez demo Python script), works the best.

Thanks, I try to look at it. Btw how is it possible to add subtitles to the request too? 

`fetchAttributes = list(struct.pack( 
             "<BIBBHBH", 0, id, 
             0,         # app id 
             1, 65535,  # title 
             2, 65535, # Subtitle ? Is it correct ? 
             3, 65535,  # message 
         )) `

**Comment by pzmarzly** [comment]
The ID seems to be correct ([NotificationAttributeIDSubtitle](https://developer.apple.com/library/archive/documentation/CoreBluetooth/Reference/AppleNotificationCenterServiceSpecification/Appendix/Appendix.html#//apple_ref/doc/uid/TP40013460-CH3-SW1)), but I think you need to add another `BH` to the struct format string.

**Comment by pzmarzly** [comment]
I've rewritten the project from scratch, so I'm closing old issues.
